File: leetcode/python/000015.3sum.py
class Solution:
	# Time complexity: O(n)
	# Space complexity: O(n) // due to Timsort
	def threeSum(self, nums: list[int]) -> list[list[int]]:
		res = []
		nums.sort()
		for i in range(len(nums) - 2):
			# Skip duplicates
			if i > 0 and nums[i-1] == nums[i]:
				continue

			cur = nums[i]
			complement = 0 - nums[i] # our target sum is 0
			j, k = i + 1, len(nums) - 1
			while j < k:
				sum = nums[j] + nums[k]
				if sum > complement:
					k -= 1
				elif sum < complement:
					j += 1
				else:
					res.append([cur, nums[j], nums[k]])
					j += 1
					# Skip duplicates
					while nums[j] == nums[j-1] and j < k:
						j += 1
		return res

File: leetcode/typescript/000015.3sum.ts
function threeSum(nums: number[]): number[][] {
  if (nums.length < 3) {
    return [];
  }

  const res: number[][] = [];

  nums.sort((a, b) => a - b);

  for (let i = 0; i <= nums.length - 3; i++) {
    // Skip duplicates
    if (i > 0 && nums[i] === nums[i - 1]) {
      continue;
    }

    // It's impossible for sum to be 0 at this point
    // because the array is sorted, and the next numbers are all >= nums[i]
    if (nums[i] > 0) {
      break;
    }

    // Two sum on sorted array
    let l = i + 1;
    let r = nums.length - 1;
    const target = 0 - nums[i];
    while (l < r) {
      // Skip duplicates
      if (l > i + 1 && nums[l] === nums[l - 1]) {
        l++;
        continue;
      }

      const sum = nums[l] + nums[r];
      if (sum > target) {
        r--;
      } else if (sum < target) {
        l++;
      } else {
        res.push([nums[i], nums[l], nums[r]]);
        l++;
        r--;
      }
    }
  }

  return res;
}
// Time complexity: O(n^2)
// Space complexity: O(1)
